Proposal for Diffserv consideration text (#151)
ietf-wg-masque/draft-ietf-masque-connect-ip: draft-ietf-masque-connect-ip.md

correctly to congestion notifications if it uses ECN. The inner traffic can
also use ECN, independently of whether it is in use on the outer connection.

## Differentiated Services Considerations {#dscp-considerations}

Tunneled IP packets can have Differentiated Services Code Points (DSCP) {{!DSCP=RFC2474}} set
in the traffic class IP header field to request a particular per-hop
behavior. If a client or IP proxy is configured as part of a Differentiated Services
domain, it MAY implement traffic differentiation based on these markings. However,
the use of HTTP can limit the possibilities for differentiated treatment of the
tunneled IP packets on the path between the IP proxying endpoints.

If tunneled packets are subject to congestion control by the outer
connection, the tunneled packets need to be treated equally regardless of

> their DSCP markings to not disrupt the congestion controller. The client or proxy MUST NOT

Reviewer: ```suggestion
their DSCP markings to not disrupt the congestion controller. In this scenario, the client or proxy MUST NOT
```

copy the DSCP field from the inner IP header to the outer IP header of the
packet carrying this packet. Instead, an application would need to use
separate connections to the proxy, one for each DSCP. Note that this
document does not define a way for requests to scope to particular DSCP
values; such support is left to future extensions.

> If tunneled packets are not subject to congestion control by the outer
> connection, and uses QUIC datagrams, the IP proxying endpoint MAY translate

Reviewer 2: ```suggestion
If tunneled packets use QUIC datagrams and are not subject to congestion control
by the outer connection, the IP proxying endpoints MAY translate
```

the DSCP field value from the tunneled traffic
to the outer IP header. IP proxying endpoints MUST NOT coalesce multiple inner
packets into the same outer packet unless they have the same DSCP marking or an

> equivalent traffic class. Note: If the client or IP proxy can copy or needs to
> translate the DSCP value depends if the tunnel ingress and egress are in the
> same differentiated service domain or different ones.

Reviewer 2: ```suggestion
equivalent traffic class. Note that the ability to translate DSCP values is
dependent on the tunnel ingress and egress belonging to the same
differentiated service domain or not.
```

correctly to congestion notifications if it uses ECN. The inner traffic can
also use ECN, independently of whether it is in use on the outer connection.

## Differentiated Services Considerations {#dscp-considerations}

Tunneled IP packets can have Differentiated Services Code Points (DSCP) {{!DSCP=RFC2474}} set
in the traffic class IP header field to request a particular per-hop

> behavior. If a client or IP proxy handling an IP proxying request is configured
> as part of a Differentiated Services domain, it MAY implement traffic differentiation. However,
> the use of HTTP can limit the possibilities for differentiated treatment of the
> tunneled IP packets on the path between the IP proxying endpoints.

Reviewer 2: ```suggestion
behavior. If a client or IP proxy is configured as part of a Differentiated Services
domain, it MAY implement traffic differentiation based on these markings. However,
the use of HTTP can limit the possibilities for differentiated treatment of the
tunneled IP packets on the path between the IP proxying endpoints.
```

> If tunneled packets are subject to congestion control by the outer 

Reviewer 2: ```suggestion
If tunneled packets are subject to congestion control by the outer
```

> connection, the tunneled packets need to be treated equally regardless of 

Reviewer 2: ```suggestion
connection, the tunneled packets need to be treated equally regardless of
```

their DSCP markings to not disrupt the congestion controller. The client or proxy MUST NOT
copy the DSCP field from the inner IP header to the outer IP header of the

> packet carrying this packet. Instead, an application intending to use IP
> proxying with different DSCP will have to establish multiple connections to a
> proxy, one per each DSCP to be used. It needs to be noted that the Connect-IP
> request per current specification does not define a way for a scoped request
> that use DSCP as a traffic selector.

Reviewer 2: ```suggestion
packet carrying this packet. Instead, an application would need to use
separate connections to the proxy, one for each DSCP. Note that this
document does not define a way for requests to scope to particular DSCP
values; such support is left to future extensions.
```

> If tunneled packets are not subject to congestion control by the outer 

Reviewer 2: ```suggestion
If tunneled packets are not subject to congestion control by the outer
```

> connection, and uses QUIC datagrams, the IP proxying endpoint MAY translate 

Reviewer 2: ```suggestion
connection, and uses QUIC datagrams, the IP proxying endpoint MAY translate
```

correctly to congestion notifications if it uses ECN. The inner traffic can
also use ECN, independently of whether it is in use on the outer connection.

> ## Diffserv Considerations {#dscp-consideration}

Reviewer: ```suggestion
## Differentiated Services Considerations {#dscp-considerations}
```

> Tunneled IP packets can have Diffserv Code Points (DSCP) {{!DSCP=RFC2474}} set

Reviewer: ```suggestion
Tunneled IP packets can have Differentiated Services Code Points (DSCP) {{!DSCP=RFC2474}} set
```

in the traffic class IP header field to request a particular per-hop
behavior. If a client or IP proxy handling an IP proxying request is configured

> as part of a diffserv domain, it MAY implement traffic differentiation. However,

Reviewer: ```suggestion
as part of a Differentiated Services domain, it MAY implement traffic differentiation. However,
```

the use of HTTP can limit the possibilities for differentiated treatment of the

> tunneled IP packets on the path between the ingress and egress.

Reviewer: ```suggestion
tunneled IP packets on the path between the IP proxying endpoints.
```

> If a client or IP proxy with a connection containing an IP Proxying request
> stream is using congestion control for that stream, regardless of whether HTTP
> datagrams or capsules are used, all traffic without of DSCP markings will be
> treated equally within that transport connection. The client or proxy MUST NOT

Reviewer: ```suggestion
If tunneled packets are subject to congestion control, they will be treated
equally regardless of their DSCP markings. IP proxying endpoints MUST NOT
```

Author: I realize that my sentence can be shorten but it actually lacks an important distinction. It states the wrong chain of dependency. Using congestion control requires one to not expose different DSCP within the connection avoid screwing up the congestion controller. So maybe formulating this as: 

If tunneled packets are subject to congestion control, they will need to be treated
equally, i.e. use the same DSCP in the tunnelling connection, regardless of the encapsulated DSCP markings.

Is that better?

Author: I committed another rewording that I hope improves things without loosing relevant details.

copy the DSCP field from the inner IP header to the outer IP header of the
packet carrying this packet. Instead, an application intending to use IP
proxying with different DSCP will have to establish multiple connections to a
proxy, one per each DSCP to be used. It needs to be noted that the Connect-IP
request per current specification does not define a way for a scoped request
that use DSCP as a traffic selector.

If a client or IP proxy with a connection containing an IP proxying request
stream and uses HTTP datagrams and disables congestion control for this stream,

> a client or proxy MAY translate the DSCP field value from the tunneled traffic

Author: Look at rewording I have commited that attempt to correct this.

correctly to congestion notifications if it uses ECN. The inner traffic can
also use ECN, independently of whether it is in use on the outer connection.

## Diffserv Considerations

> The tunneled IP traffic can have a Diffserv Code Point (DSCP) {{!DSCP=RFC2474}}

Reviewer 2: ```suggestion
Tunneled IP packets can have Diffserv Code Points (DSCP) {{!DSCP=RFC2474}}
```

> set in the traffic class IP header field to request a particular per hop

Reviewer 2: ```suggestion
set in the traffic class IP header field to request a particular per-hop
```

> behavior. The client or IP proxy handling an IP proxying request if part of the
> differv domain and configured with the domains mapping MAY implement traffic

Reviewer 2: ```suggestion
behavior. If a client or IP proxy handling an IP proxying request is configured as part
of a diffserv domain, it MAY implement traffic
```

differentiation. However, the use of HTTP can limit the possibilities for
differentiated treatment of the tunneled IP packets on the path between the
ingress and egress.

If a client or IP proxy with a connection containing an IP Proxying request

> stream are using congestion control for that stream, independent if datagram or

Reviewer 2: ```suggestion
stream is using congestion control for that stream, regardless of whether HTTP datagrams or
```

> capsules are used, all traffic independent of DSCP marking will be treated

Reviewer 2: ```suggestion
capsules are used, all traffic without of DSCP markings will be treated
```

equally within that transport connection. The client or proxy MUST NOT copy the
DSCP field from the inner IP header to the outer IP header of the packet

> carrying this packet. Instead an application intending to use IP proxying with

Reviewer 2: ```suggestion
carrying this packet. Instead, an application intending to use IP proxying with
```

> different DSCP will have to establish multiple transport connections, one per

Reviewer 2: ```suggestion
different DSCP will have to establish multiple connections to a proxy, one per
```

> each DSCP to be used, and requesting proxying of these flows specifically. It

Reviewer 2: ```suggestion
each DSCP to be used. It
```

need to be noted that as this specification does not allow DSCP as a selector

> for the traffic matching the Connect-IP request.

Reviewer 2: I can't quite parse this sentence.

Author: So I have pushed a text change that reword that to: 

It needs to be noted that the Connect-IP request per current specification does not define a way for a scoped request that use DSCP as a traffic selector.

Is that clearer?

If a client or IP proxy with a connection containing an IP proxying request

> stream and uses datagram and disables congestion control for this stream a

Reviewer 2: ```suggestion
stream and uses HTTP datagrams and disables congestion control for this stream, a
```

client or proxy MAY translate the DSCP field value from the tunneled traffic to
the outer IP header for the QUIC packet containing the datagram. Coalescing of
multiple datagrams in one QUIC packet MAY only be done if the the IP packets
encapsulated in each datagram have the same DSCP or an equivalent traffic
class. Note: If the client or IP proxy can copy or needs to translate the DSCP
value depends if the tunnel ingress and egress are in the same differentiated

> service domain or different ones.

Reviewer 3: Not sure if that last part of the advise is appropriate. Usually there is no tight control about which frames go into with packet and a packet containing a datagram frames may also carry other control frames. Also making packets with different DSCP may then also expose something about the content of the packet that you potentially don't want to expose.

Author: So your point is excellent. One clearly need strict control between the datagram encapsulated inner packet and its mapping to a particular outer if you are going to do this. 

Secondly, you are correct that one should add a bit of security consideration to this also to warn about the fact that one likely exposes flow structure. And that one should not do this if one have issues with exposing this.

Author: I have now added security consideration text to this PR.

{{?ROUTING-HDR=RFC5095}}), implementers need to follow the latest guidance
regarding handling of IPv6 extension headers.

> Client or Proxies that support marking DSCP through to encapsulating
> datagram carrying packets, see {{dscp-consideration}}, needs to be
> aware that this will expose end-to-end flow level information to an
> on-path observer between the client and proxy. The marking through of
> DSCP will result in that one common aggregate over the connection
> becomes multiple aggregates, one per DSCP used. Potentially exposing a
> single end-to-end flow. Thus, if the goal of IP proxying over HTTP is
> privacy related then marking DSCP through is not recommended.

Reviewer: ```suggestion
Transferring DSCP markings from inner to outer packets (see {{dscp-considerations}})
exposes end-to-end flow level information to an on-path observer between the IP
proxying endpoints. This can potentially expose a single end-to-end flow. Because of
this, such use of DSCP in privacy-sensitive contexts is NOT RECOMMENDED.
```
